### ISSUE TITLE
mapNonNull and filterNonNull leave nullable types intact

### ISSUE THREAD

**Issue by viert** [apple/pkl-intellij#2]
Hi team,

`mapNonNull` and `filterNonNull` should presumably change the resulting type from `Listing<Elem?>` to `Listing<Elem>` as the result is guaranteed to not have nulls. The warning is misleading and requires a redundant `as Listing<Elem>` extra.
 
<img width="816" alt="Fullscreen_05_02_2024__17_00" src="https://github.com/apple/pkl-intellij/assets/1122152/dfdeadac-f1f2-4879-ae6f-b3200a0d6c07">

**Comment by holzensp** [comment]
Yes, this is a known issue. For `mapNonNull`, this should actually be doable. For `filterNonNull` there's a Hard Problem™ to solve. What you're seeing here is that _static_ type analysis that only happens in the plugin (not the interpreter). It does this by taking the type signature as specified.

For `mapNonNull`, we could change the type from
```
mapNonNull<Result>(transform: (Element) -> Result): Collection<Result(this != null)>
```
to
```
mapNonNull<Result>(transform: (Element) -> Result?): Collection<Result>
```
(and similar for `List` and `Set`)

The **downside of this** is that it becomes a type mismatch when the plugin can't explicitly decide that your `transform` results in a nullable `Result` (this happens disappointingly more often that you might think). That would mean it explicitly requires `as MyType?` asserting.

For `filterNonNull`, the only route I can see is special-casing it in the plugin, because the type is
```
filterNonNull(): Collection<Element(this != null)>
```
(and similar for `List` and `Set`)

To change use the specified type signatures for that, Pkl would need to support something like [_Generic Where Clauses_ ](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/generics/#Generic-Where-Clauses) in Swift (similar; _conditional methods_ in C#), because, as you can see, the definition of `Element` lives on the `class`, not on the method's signature.
